atox/src/main/kotlin/ui/AvatarImageView.kt

```diff
@@ -0,0 +1,97 @@
+package ltd.evilcorp.atox.ui
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.LayoutInflater
+import android.widget.FrameLayout
+import androidx.constraintlayout.widget.ConstraintLayout
+import kotlin.math.pow
+import kotlin.math.sqrt
+import ltd.evilcorp.atox.R
+import ltd.evilcorp.atox.databinding.AvatarImageLayoutBinding
+
+private const val STATUS_INDICATOR_SIZE_RATIO_WITH_AVATAR = 12f / 50
+
+class AvatarImageView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
+    FrameLayout(context, attrs, defStyleAttr) {
+
+    private val binding = AvatarImageLayoutBinding.inflate(LayoutInflater.from(context), this, true)
+
+    val avatarImage get() = binding.avatarImage
+
+    val statusIndicator get() = binding.statusIndicator
+
+    var size: Int = 0
```

Comment:
The behaviour of `size` isn't very intuitive to work with compared to `layout_width`/`layout_height`. It's especially obvious when things aren't a fixed size as can be seen in the hack required to dynamically size the call background in 93a4aa1fddcc1011ecb15bb6b6be0c9ce9e2c0a1. Normally you'd use something like weights instead of having to register/unregister listeners, and using weights continue working if even if things change size from e.g. being animated.

Is it possible to continue using `layout_{width,height}` for sizing this?

Reply:
It's possible but you would still need the layout listener for getting the height/width of the view for the use of AvatarFactory.

Reply:
We can also just change the `assignInto` function in AvatarFactory to execute its code in a runnable that is passed to the AvatarImageView.post method like is suggested in solution 2 in the link above.

Reply:
And that way we can also return the `ImageView` argument to `assignInto` instead of changing it to `AvatarImageView` while still removing the `size` argument.

Reply:
It was hard... But I succeeded to do it with the `doOnLayout` kotlin extension (which I should have also used it instead of the layout listener). However, the `assignInto` method in `AvatarFactory` still needs a container like `AvatarImageView`. How about moving that method to `AvatarImageView` and deleting `AvatarFactory`?

Reply:
Oh the extension just uses what I did which is listen for layout change and then remove the listener. But that's how you do it and that's what the `androidx` library provides.

Reply:
>It was hard... But I succeeded to do it only with [...]

Great job!

>How about moving that method to AvatarImageView and deleting AvatarFactory? We can also move the statusIndicator color [...]

Yeah, sure! Sounds like a good idea! We could also just move AvatarFactory into the same file as AvatarImageView and make it a private class to keep UI separated from avatar creation, whatever seems nicest when you mess around with it. :P

>Oh the extension just uses what I did which is [...]

Yeah, fair enough if that's the way it's done. It just seemed like a pretty odd way of doing it, but I have also not built any custom XML bits myself yet. :P If we find a nicer way of doing things later it's always possible to change it then.

Reply:
Thank you. Yeah I also considered making it a private class. I'll see what feels better.
The extension is a very nice way to do it in terms of simplicity. But under the hood it's just an inline function that uses what I did but also checks first if the layout has been created already (in this case it doesn't create a listener and just execute the lambda right away).